README
WWW-YouTube-Info-Simple version 0.07
====================================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Carp;
  Data::Dumper;
  WWW::YouTube::Info

COPYRIGHT AND LICENCE

Copyright (C) 2011 by east

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension WWW::YouTube::Info::Simple.

0.01  Mon Aug 23 11:26:42 2010
	- original version; created by h2xs 1.23 with options
		-AX WWW::YouTube::Info::Simple
0.02  Fri Aug 27 13:03:38 2010
	- spelling and formating of POD
0.03  Fri Aug 27 15:45:37 2010
	- fixed potential collision at export
0.04  Wed Sep  1 10:52:24 2010
	- upgrades at prereq
0.05  Sun Sep  5 00:18:01 2010
	- fixed encoding at get_keywords
0.06  Sat Jul 23 21:50:45 2011
	- added method get_conn
	- added tests
0.07  Thu Aug 11 13:57:09 2011
	- added fallbacks to get_url(), get_resolution()
		fmt_url_map, fmt_map seem to be lost more and more



examples/yt.pl
#!/usr/bin/perl

# CLI example usage of WWW::YouTube::Info::Simple
# to directly download YouTube videos
# with means of wget
# by a given URL or VIDEO_ID.
# e.g.: ./yt.pl http://youtube.com/watch?v=foobar
# e.g.: ./yt.pl http://www.youtube.com/v/foobar
# e.g.: ./yt.pl foobar

# change /usr/bin/wget to the appropriate path/bin
# and use it at your own risk.

use strict;
use warnings;

use Data::Dumper;
use WWW::YouTube::Info::Simple;
use LWP::Simple;
use File::stat;

my $val = $ARGV[0] || die "usage: $0 '<yturl|id>'\n";

# get VIDEO_ID from $ARGV[0]
my $id = get_id($val) or die "no VIDEO_ID found at $val!";

my $yt = WWW::YouTube::Info::Simple->new($id);

# check status
my $info = $yt->get_info();
die "status ne 'ok' ('$info->{status}')!" if $info->{status} ne 'ok';

# get details
my $url   = $yt->get_url();
my $res   = $yt->get_resolution();
my $title = $yt->get_title();

# title is to be used as filename
# hence, remove slashes
$title =~ s/\///g;

# show qualities to choose from
print "Sources found:\n";
foreach my $key ( keys %$res ) {
  printf "Quality: %-2s [%s]\n", $key, $res->{$key};
}

# ask what quality to choose
my $q;
while ( 1 ) {
  print "Enter quality of choice: ";
  my $in = <STDIN>;
  chomp $in;
  # check whether choice exists
  $q = $in and last if exists $url->{$in};
  print "Choice has no URL!\n";
}

# get remote filetype/fileextension
my (
  $content_type,
  $document_length,
  $modified_time,
  $expires,
  $server
) = head($url->{$q});

my $ext;
if ( $content_type ) {
  # .. by header info
  $ext = '.flv' if ( $content_type =~ m/flv/ );
  $ext = '.mp4' if ( $content_type =~ m/mp4/ );
  die "no known content_type found: ($content_type)!" unless $ext;
}
else {
  die "no content_type found!";
}
my $file = $title.$ext;

# check whether file exists?
if ( -e $file ) {
  # get remote filesize
  # .. by header info
  my $url_size_mb  = $document_length ? sprintf "%.3f", $document_length/1048576 : 'n.a.';
  # get local filesize
  my $file_size_mb = sprintf "%.3f", stat($file)->size/1048576;

  print "Filename exists! [$file]\n";
  print "Local:  $file_size_mb MB\n";
  print "Remote: $url_size_mb MB\n";
  print "Overwrite? [y/N]: ";

  # ask what to do
  my $in = <STDIN>;
  chomp $in;
  unless ( $in =~ /^(y|yes|Y|Yes)$/ ) {
    print "Aborting.\n";
    exit(0);
  }
  print "\n";
}

# download
print "Downloading ..\n";
system('/usr/bin/wget', "-O$file", $url->{$q});

exit(0);


sub get_id {
  my $val = shift;

  my $uri   = URI->new($val);
  my $path  = $uri->path();
  my $query = $uri->query();

  my $id;
  # http://youtube.com/watch?v=foobar
  if ( $path =~ /\/watch/ ) {
    $id = $query;
    $id =~ s/^v=//;
  }
  # http://www.youtube.com/v/foobar
  elsif ( $path =~ /\/v\// ) {
    $id = $path;
    $id =~ s/\/v\///;
  }
  # foobar
  elsif ( $path !~ /\// ) {
    $id = $path;
  }

  return $id;
}

1;



lib/WWW/YouTube/Info/Simple.pm
package WWW::YouTube::Info::Simple;

use 5.008;
use strict;
use warnings;

require Exporter;
require WWW::YouTube::Info;

our @ISA = qw(
  Exporter
  WWW::YouTube::Info
);

our @EXPORT = qw(
);

our $VERSION = '0.07';

use Carp;
use Data::Dumper;

=head1 NAME

WWW::YouTube::Info::Simple - simple interface to WWW::YouTube::Info

=head1 SYNOPSIS

=head2 Perhaps a little code snippet?

  #!/usr/bin/perl
  
  use strict;
  use warnings;
  
  use WWW::YouTube::Info::Simple;
  
  # id taken from YouTube video URL
  my $id = 'foobar';
  
  my $yt = WWW::YouTube::Info::Simple->new($id);
  
  # hash reference holds values gained via http://youtube.com/get_video_info?video_id=foobar
  my $info = $yt->get_info();
  # this is basically an inheritance to get_info() at WWW::YouTube::Info
  # $info->{title}          # e.g.: Foo+bar+-+%27Foobar%27
  # $info->{author}         # e.g.: foobar
  # $info->{keywords}       # e.g.: Foo%2Cbar%2CFoobar
  # $info->{length_seconds} # e.g.: 60
  # $info->{fmt_map}        # e.g.: 22%2F1280x720%2F9%2F0%2F115%2C35%2F854x480%2F9%2F0%2F115%2C34%2F640x360%2F9%2 ..
  # $info->{fmt_url_map}    # e.g.: 22%7Chttp%3A%2F%2Fv14.lscache1.c.youtube.com%2Fvideoplayback%3Fip%3D131.0.0.0 ..
  # $info->{fmt_stream_map} # e.g.: 22%7Chttp%3A%2F%2Fv14.lscache1.c.youtube.com%2Fvideoplayback%3Fip%3D131.0.0.0 ..
  
  # array reference holds values keywords
  my $keys = $yt->get_keywords();
  # $keys->[0] # e.g.: Foo
  # $keys->[1] # e.g.: bar
  # ..
  
  # hash reference holds values quality -> resolution
  my $res = $yt->get_resolution();
  # $res->{35} # e.g.: 854x480
  # $res->{22} # e.g.: 1280x720
  # ..
  
  # URL and masquerading decoded title
  my $title = $yt->get_title(); # e.g.: Foo bar - 'Foobar'
  
  # hash reference holds values quality -> url
  my $url = $yt->get_url();
  # $url->{35} e.g.: http://v14.lscache1.c.youtube.com/videoplayback?ip=131.0.0.0 ..
  # $url->{22} e.g.: http://v14.lscache1.c.youtube.com/videoplayback?ip=131.0.0.0 ..
  # ..
  
  # URL decoded RTMPE URL
  my $conn = $yt->get_conn(); # e.g.: rtmpe://cp59009.edgefcs.net/youtube?auth=daEcaboc8dvawbcbxazdobDcZajcDdgcfae ..
  
  # Remark:
  # You might want to check $info->{status} before further workout,
  # as some videos have copyright issues indicated, for instance, by
  # $info->{status} ne 'ok'.

=head1 DESCRIPTION

I guess its pretty much self-explanatory ..

=head1 METHODS

=cut

=head2 get_keywords

Returns undef if status ne 'ok'.
Croaks if not available.

=cut

sub get_keywords {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  my $keywords = $self->{info}->{'keywords'};
  croak "no keywords found!" unless $keywords;

  my @keywords_parts = split /%2C/, $keywords;
  foreach my $item ( @keywords_parts ) {
    next unless $item;
    $item = _url_decode($item);
    push @{$self->{keywords}}, $item;
  }

  return $self->{keywords};
}

=head2 get_resolution

Returns undef if status ne 'ok'.
Croaks if not available.

=cut

sub get_resolution {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  # quality and resolution
  my $fmt_map = $self->{info}->{'fmt_map'};
  unless ( $fmt_map ) {
    # fallback to fmt_list
    # as fmt_map doesnt't seem to be supported any more
    croak "no resolutions found!" unless $self->_fmt_list();
  }
  else {
    # process fmt_map
    my @fmt_map_parts = split /%2F9%2F0%2F115%2C/, $fmt_map;
    foreach my $item ( @fmt_map_parts ) {
      my ($quality, $resolution) = split /%2F/, $item;
      next unless $quality and $resolution;
      $self->{resolution}->{$quality} = $resolution;
    }
  }

  return $self->{resolution};
}

=head2 get_title

Returns undef if status ne 'ok'.
Croaks if not available.

=cut

sub get_title {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  my $title = $self->{info}->{'title'};
  croak "no title found!" unless $title;

  $title = _url_decode($title);
  $title =~ s/\+/ /g;
  $self->{title} = $title;

  return $self->{title};
}

=head2 get_url

Returns undef if status ne 'ok'.
Croaks if not available.

  use WWW::YouTube::Info::Simple;
  
  # id taken from YouTube video URL
  my $id = 'foobar';
  
  my $yt = WWW::YouTube::Info::Simple->new($id);
  
  # hash reference holds values quality -> url
  my $url = $yt->get_url();
  # $url->{35} e.g.: http://v14.lscache1.c.youtube.com/videoplayback?ip=131.0.0.0 ..
  # $url->{22} e.g.: http://v14.lscache1.c.youtube.com/videoplayback?ip=131.0.0.0 ..
  # ..

YouTube videos can be downloaded in given qualities by means of these URLs and the usual suspects (C<wget>, ..).

=cut

sub get_url {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  # quality and URL
  my $fmt_url_map = $self->{info}->{'fmt_url_map'};
  unless ( $fmt_url_map ) {
    # fallback to url_encoded_fmt_stream_map
    # as fmt_url_map doesnt't seem to be supported any more
    croak "no URLs found!" unless $self->_url_encoded_fmt_stream_map();
  }
  else {
    # process fmt_url_map
    my @fmt_url_map_parts = split /%2C/, $fmt_url_map;
    foreach my $item ( @fmt_url_map_parts ) {
      my ($quality, $url) = split /%7C/, $item;
      next unless $quality and $url;
      $url = _url_decode($url);
      $self->{url}->{$quality} = $url;
    }
  }

  return $self->{url};
}

=head2 get_conn

Returns undef if status ne 'ok'.
Croaks if not available.

  use WWW::YouTube::Info::Simple;
  
  # id taken from YouTube video URL
  my $id = 'foobar';
  
  my $yt = WWW::YouTube::Info::Simple->new($id);
  
  # URL decoded RTMPE URL
  my $conn = $yt->get_conn(); # e.g.: rtmpe://cp59009.edgefcs.net/youtube?auth=daEcaboc8dvawbcbxazdobDcZajcDdgcfae ..

A YouTube RTMPE stream can be accessed via this URL and downloaded by
means of the usual suspects (C<rtmpdump>, ..).
The URL looses its validity after approx. 30 seconds (experimental value).
Gathering a fresh RTMPE URL regarding the same VIDEO_ID and the
C<rtmpdump .. --resume> capability might circumvent this inconvenience.

=cut

sub get_conn {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  my $conn = $self->{info}->{'conn'};
  croak "no conn found!" unless $conn;

  $self->{conn} = _url_decode($conn);

  return $self->{conn};
}


sub _url_encoded_fmt_stream_map {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  # quality and URL
  my $url_encoded_fmt_stream_map = $self->{info}->{'url_encoded_fmt_stream_map'};
  return unless $url_encoded_fmt_stream_map;

  my @url_encoded_fmt_stream_map_parts = split /%2C/, $url_encoded_fmt_stream_map;
  foreach my $item ( @url_encoded_fmt_stream_map_parts ) {
    my $url = _url_decode($item);
    $url =~ s/.*url=(.*)&fallback_host=.*/$1/;
    $url = _url_decode($url);
    (my $quality = $url) =~ s/.*&itag=(\d+)&.*/$1/;
    next unless $quality and $url;
    $self->{url}->{$quality} = $url;
  }

  return $self->{url};
}

sub _fmt_list {
  my ($self) = @_;

  $self->get_info() unless exists($self->{info});
  return if ( $self->{info}->{status} ne 'ok' );

  # quality and resolution
  my $fmt_list = $self->{info}->{'fmt_list'};
  return unless $fmt_list;

  my @fmt_list_parts = split /%2C/, $fmt_list;
  foreach my $item ( @fmt_list_parts ) {
    my ($quality, $resolution, @rest) = split /%2F/, $item;
    next unless $quality and $resolution;
    $self->{resolution}->{$quality} = $resolution;
  }

  return $self->{resolution};
}

sub _url_encode {
  my $string = shift;

  # URLencode
  $string =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;

  return $string;
}

sub _url_decode {
  my $string = shift;

  # URLdecode
  $string =~ s/\%([A-Fa-f0-9]{2})/pack('C', hex($1))/seg;

  return $string;
}

1;

__END__

=head1 SEE ALSO

You might want to have a look at the C<./examples> folder within this
distribution, or at L<WWW::YouTube::Info>.

=head1 HINTS

Searching the internet regarding 'fmt_url_map', 'url_encoded_fmt_stream_map'
and/or 'get_video_info' might gain hints/information to improve
L<WWW::YouTube::Info> and L<WWW::YouTube::Info::Simple> as well.

=head1 BUGS

Please report bugs and/or feature requests to
C<bug-WWW-YouTube-Info-Simple at rt.cpan.org>,
alternatively by means of the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-YouTube-Info-Simple>.

=head1 AUTHOR

east E<lt>east@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by east

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.

=cut

